### PR TITLE
fix: check no_changes on promote content

### PR DESCRIPTION
**Pull request by rephus**
Closes:  https://github.com/lightdash/lightdash/issues/10370

### Description:

Chart promote with space name change but no new chart version

Screenshot from 2024-06-13 12-10-24

Dashboard promote without chart changes 
Screenshot from 2024-06-13 11-06-26

No changes on chart promote 

Screenshot from 2024-06-13 10-51-12

### Limitations

- Dashboard don't have updated_at so we assume they are always updated (existing behaviour) 
- We only promote for charts if promotedChart.updatedAt > upstreamChart.updatedAt or name/description changes. 

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

File: packages/backend/src/services/PromoteService/PromoteService.ts [hunk 1]
```diff
@@ -811,16 +847,24 @@ export class PromoteService extends BaseService {
         };
     }
 
-    async createNewSpaces(
+    async upsertSpaces(
```
Comment (rephus) [inline packages/backend/src/services/PromoteService/PromoteService.ts:850]: now we can detect `name` changes in spaces, so we can also   update spaces if needed

File: packages/backend/src/services/PromoteService/PromoteService.ts [hunk 2]
```diff
@@ -437,6 +437,45 @@ export class PromoteService extends BaseService {
         );
     }
 
+    private static isSpaceUpdated(
```
Comment (rephus) [inline packages/backend/src/services/PromoteService/PromoteService.ts:440]: extracted updated logic on static method for spaces/chart/dashboard

File: packages/backend/src/services/PromoteService/PromoteService.ts [hunk 3]
```diff
@@ -882,60 +931,96 @@ export class PromoteService extends BaseService {
...
 
         return {
-            spaces: [...existingSpaces, ...newSpaceChanges],
+            spaces: allSpaces,
             dashboards: updateDashboardWithSpace,
             charts: updateChartsWithSpace,
         };
     }
 
+    static getChartChange(
```
Comment (rephus) [inline packages/backend/src/services/PromoteService/PromoteService.ts:940]: extracted and unified logic to be used in `promoteChart` and `promoteDashboard`

File: packages/backend/src/services/PromoteService/PromoteService.ts [hunk 4]
```diff
@@ -882,60 +931,96 @@ export class PromoteService extends BaseService {
...
+                spaceUuid:
+                    upstreamChart.space?.uuid || promotedChart.space.uuid, // set the new space uuid after creation
+                projectUuid: promotedChart.projectUuid,
+                spaceSlug: promotedChart.space?.slug,
+                oldUuid: promotedChart.chart.uuid,
+            },
+        };
+    }
+
+    static getSpaceChange(
```
Comment (rephus) [inline packages/backend/src/services/PromoteService/PromoteService.ts:978]: extracted and unified logic to be used in `promoteChart` and `promoteDashboard`

File: packages/backend/src/services/PromoteService/PromoteService.ts [hunk 5]
```diff
@@ -1053,46 +1121,13 @@ export class PromoteService extends BaseService {
...
-                        oldUuid: promotedChart.chart.uuid,
-                    },
-                };
-            },
+            ({ promotedChart, upstreamChart }) =>
+                PromoteService.getChartChange(promotedChart, upstreamChart),
         );
 
         // TODO return charts within dashboards that are going to be deleted after the promotion
+        // For this we'll need to get all the tiles for the upstreamDashboard and compare against the promotedDashboard.tiles
```
Comment (rephus) [inline packages/backend/src/services/PromoteService/PromoteService.ts:1129]: I tried to implement this, but we are also missing upstreamDashboard.tiles, so not for now

**Review by ZeRego — APPROVED** [review]

**Pull request merged**